src/ua_types_encoding_json.c

```diff
@@ -0,0 +1,5101 @@
+                    ret |= CALC_DIRECT(&jsonEncodingField, Byte);
+                }
+                ret |= calcWriteKey(ctx, "Body", commaNeeded);
+                ret |= CALC_DIRECT(&src->content.encoded.body, String);
+                break;
+            }
+            case UA_EXTENSIONOBJECT_ENCODED_XML:
+            {
+                if(useReversible){
+                    UA_Byte jsonEncodingField = 2;
+                    ret |= calcWriteKey(ctx, "Encoding", commaNeeded);
+                    commaNeeded = true;
+                    ret |= CALC_DIRECT(&jsonEncodingField, Byte);
+                }
+                ret |= calcWriteKey(ctx, "Body", commaNeeded);
+                ret |= CALC_DIRECT(&src->content.encoded.body, String);
+                break;
+            }
+            default:
+                ret = UA_STATUSCODE_BADINTERNALERROR;
+        }
+
+        ret |= ADDJSONCHAR(ObjEnd);
+    }else{ 
+        
+        /* Cannot encode with no type description */
+        if (!src->content.decoded.type)
+            return UA_STATUSCODE_BADENCODINGERROR;
+
+        if(!src->content.decoded.data){
+            return calcWriteNull(ctx);
+        }
+        
+        UA_NodeId typeId = src->content.decoded.type->typeId;
+        if (typeId.identifierType != UA_NODEIDTYPE_NUMERIC)
+            return UA_STATUSCODE_BADENCODINGERROR;
+        
+        if (useReversible) {
+            /* REVERSIBLE */
+            ret |= ADDJSONCHAR(ObjStart);
+
+            ret |= calcWriteKey(ctx, "TypeId", UA_FALSE);
+            ret |= CALC_DIRECT(&typeId, NodeId);
+            if (ret != UA_STATUSCODE_GOOD)
+                return ret;
+            
+            /* Encoding field is omitted if the value is 0. */
+            const UA_DataType *contentType = src->content.decoded.type;
+
+            /* Encode the content */
+            ret |= calcWriteKey(ctx, "Body", UA_TRUE);
+            ret |= calcJsonInternal(src->content.decoded.data, contentType, ctx, useReversible);
+            if (ret != UA_STATUSCODE_GOOD)
+                return ret;
+
+            ret |= ADDJSONCHAR(ObjEnd);
+        } else {
+            /* -NON-REVERSIBLE
+             * For the non-reversible form, ExtensionObject values 
+             * shall be encoded as a JSON object containing only the 
+             * value of the Body field. The TypeId and Encoding fields are dropped.
+             * 
+             * TODO: "Body" key in the ExtensionObject?
+             */
+            ret |= ADDJSONCHAR(ObjStart);
+            const UA_DataType *contentType = src->content.decoded.type;
+            ret |= calcWriteKey(ctx, "Body", UA_FALSE);
+            ret |= calcJsonInternal(src->content.decoded.data, contentType, ctx, useReversible);
+            ret |= ADDJSONCHAR(ObjEnd);
+        }
+    }
+    return ret;
+}
+
+static status
+Variant_calcJsonWrapExtensionObject(const UA_Variant *src, const bool isArray, CtxJson *ctx, UA_Boolean useReversible) {
```

Comment:
That line is too long. I'm in the habit of splitting the editor window vertically.
We have no fixed rule. But maybe we can keep to <120 chars per line.